url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/util/IO.java

```diff
@@ -0,0 +1,131 @@
+import java.io.Reader;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringTokenizer;
+
+public class IO
+{
+    public static String toString(Reader r) throws IOException {
+        StringWriter sw = new StringWriter();
+        copy(r, sw, true);
+        return sw.toString();
+    }
+
+    public static String toString(InputStream is) throws IOException {
+        return toString(new InputStreamReader(is));
+    }
+
+    public static String toString(File f) throws IOException {
+        return toString(new FileReader(f));
+    }
+
+    public static void copy(Reader r, Writer w, boolean close)
+            throws IOException {
+        char[] buf = new char[4096];
+        int len = 0;
+        while ((len = r.read(buf)) > 0) {
+            w.write(buf, 0, len);
+        }
+        if (close) {
+            r.close();
+            w.close();
+        }
+    }
+
+    public static void copy(InputStream r, OutputStream w, boolean close)
+            throws IOException {
+        byte[] buf = new byte[4096];
+        int len = 0;
+        while ((len = r.read(buf)) > 0) {
+            w.write(buf, 0, len);
+        }
+        if (close) {
+            r.close();
+            w.close();
+        }
+    }
+
+    public static String getExtension(File f) {
+        if (f == null)
+            return null;
+        String n = f.getName();
+        int idx = n.lastIndexOf('.');
+        if (idx == -1) {
+            return null;
+        }
+        return n.substring(idx + 1);
+    }
+
+    public static String removeExtension(File f) {
+        if (f == null)
+            return null;
+        String n = f.getName();
+        int idx = n.lastIndexOf('.');
+        if (idx == -1)
+            return n;
+        return n.substring(0, idx);
+    }
+
+    public static byte[] toBytes(File file) throws IOException {
+        byte[] b = new byte[(int) file.length()];
+        FileInputStream fis = new FileInputStream(file);
+        fis.read(b);
+        fis.close();
+        return b;
+    }
+
+    public static BufferedReader openUrl(String url) throws IOException {
+        URL u = new URL(url);
+        return new BufferedReader(new InputStreamReader(u.openStream()));
+    }
+
+    public static String[] readLines(Reader r) throws IOException {
+        List l = new ArrayList();
+        BufferedReader br = new BufferedReader(r);
+        String line = null;
+        while ((line = br.readLine()) != null) {
+            l.add(line);
+        }
+        return (String[]) l.toArray(new String[0]);
+    }
+
+    public static String[] readLines(InputStream is) throws IOException {
+        return readLines(new InputStreamReader(is));
+    }
+
+    public static String[] readLines(Class clazz, String resource)
+            throws IOException {
+        return readLines(new InputStreamReader(clazz
+                .getResourceAsStream(resource)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *InputStreamReader*. Currently, there are execution paths that do not contain closure statements, for example, when IO.readLines throws an exception. Ensure the object returned by *InputStreamReader()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).